packages/app/obojobo-repository/server/routes/api.js

```diff
@@ -186,10 +188,39 @@ router
 			}
 
 			const oldDraft = await Draft.fetchById(draftId)
+
+			// gather all of the node IDs in the document and determine a new ID to replace each with
+			const idsForChange = {}
+
+			// this should never not exist, but just in case
+			if (oldDraft.nodesById) {
+				// only generate a replacement for ids that are UUIDs, not custom
+				const uuidRegex = /[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i
+				oldDraft.nodesById.forEach((node, key) => {
+					if (key && uuidRegex.test(key)) idsForChange[key] = uuid()
+				})
+			}
+
+			// now convert the updated document to an object for use
 			const draftObject = oldDraft.root.toObject()
+
 			const newTitle = req.body.title ? req.body.title : draftObject.content.title + ' Copy'
 			draftObject.content.title = newTitle
-			const newDraft = await Draft.createWithContent(userId, draftObject)
+
+			// convert the object to a JSON string so we can swap out all the old IDs with the new ones
+			let draftString = JSON.stringify(draftObject)
+
+			// globally replace each old ID with the equivalent new ID
+			// this should replace node IDs as well as action trigger references to those node IDs
+			for (const [oldId, newId] of Object.entries(idsForChange)) {
+				// this works, but there may be a more efficient way of doing it
```

Comment:
This is a very elegant solution... Changing keys while retaining values is a pain, while string replacing is simple and relatively fast. The only real concern is making sure you're replacing globally instead of the first instance only, which you've done below AND made sure was included in your test case.